pandas/core/series.py

```diff
@@ -2635,30 +2634,29 @@ def combine(self, other, func, fill_value=None):
                 rv = other.get(idx, fill_value)
                 with np.errstate(all='ignore'):
                     new_values.append(func(lv, rv))
+            new_dtype = type(func(lv, rv))
         else:
             # Assume that other is a scalar, so apply the function for
             # each element in the Series
             new_index = self.index
             with np.errstate(all='ignore'):
                 new_values = [func(lv, other) for lv in self._values]
             new_name = self.name
+            new_dtype = type(func(self._values[0], other))
 
         if is_categorical_dtype(self.values):
             pass
-        elif is_bool(new_values[0]) and not is_sparse(self.values):
-            pass
         elif is_extension_array_dtype(self.values):
             # The function can return something of any type, so check
             # if the type is compatible with the calling EA.
             try:
-                new_values = self._values._from_sequence(new_values)
+                new_values = self._values._from_sequence(new_values, dtype=new_dtype)
```

Comment:
you might be able to do:
```
from pandas.core.dtypes.cast import infer_dtype_from
....
new_dtype = infer_dtype_from(new_values)
```

Reply:
@jreback, is there any other way? If `new_values` contains NaNs this will return `np.float64`.

Reply:
ahh i c, ok does your current approach pass?

Reply:
It is as long as the first/last value is not nan either:) I could possibly iterate over `new_values` to find first non-nan value and use it's dtype, which however is O(n) AND would still fail to infer if every value is nan (this we could fallback to the original dtype, I guess).

Reply:
See my two latest commits (Sorry, first time accidentally committed and pushed incomplete changes(